src/Engine/ProtoCore/DSASM/Heap.cs

```diff
@@ -203,25 +192,83 @@ public int GetHashCode(StackValue value)
         }
     }
 
+    /// <summary>
+    /// String table to store all DS strings. 
+    /// </summary>
+    internal class StringTable
+    {
+        private Dictionary<string, int> stringToPointerTable;
+        private Dictionary<int, string> pointerToStringTable;
+
+        internal StringTable()
+        {
+            stringToPointerTable = new Dictionary<string, int>();
+            pointerToStringTable = new Dictionary<int, string>();
+        }
+
+        /// <summary>
+        /// Add string to the string table. 
+        /// </summary>
+        /// <param name="pointer">The index of HeapElement that represents the string</param>
+        /// <param name="s"></param>
+        internal void AddString(int pointer, string s)
+        {
+            stringToPointerTable[s] = pointer;
+            pointerToStringTable[pointer] = s;
+        }
+
+        /// <summary>
+        /// Get string from the string table.
+        /// </summary>
+        /// <param name="pointer">The index of HeapElement that represents the string</param>
+        /// <param name="s"></param>
+        /// <returns></returns>
+        internal bool TryGetString(int pointer, out string s)
+        {
+            return pointerToStringTable.TryGetValue(pointer, out s);
+        }
+
+        /// <summary>
+        /// Get the index of HeapElement that represents the string
+        /// </summary>
+        /// <param name="s"></param>
+        /// <param name="pointer"></param>
+        /// <returns></returns>
+        internal bool TryGetPointer(string s, out int pointer)
+        {
+            return stringToPointerTable.TryGetValue(s, out pointer);
+        }
+
+        internal bool TryRemoveString(int pointer)
+        {
+            string stringToBeRemoved = null;
+            if (!pointerToStringTable.TryGetValue(pointer, out stringToBeRemoved))
+                return false;
+
+            pointerToStringTable.Remove(pointer);
+            stringToPointerTable.Remove(stringToBeRemoved);
+            return true;
+        }
+    }
+
     public class Heap
     {
         private readonly List<int> freeList = new List<int>();
         private readonly List<HeapElement> heapElements = new List<HeapElement>();
+        private HashSet<int> fixedHeapElements = new HashSet<int>(); 
+        private StringTable stringTable = new StringTable();
```

Comment:
Is there a reason for the string table to live in the Heap?
1. I'm thinking for it to be moved to CompileCore (When ive done implementing it)
2. After compilation, copy it to DSExecutable.
The reason is that its consistent with the function,class and symbol tables.
If you agree, this is to be done down the line, not in this PR (I will be responsible for that)